modules/internal/ChapelArray.chpl

```diff
@@ -2923,11 +2923,63 @@ module ChapelArray {
     }
   }
 
+  // The following are the historical reshape() procedures that rely
+  // on copying the array's elements, proposed to be replaced by the
+  // pre-edition variants that follow
+
   /* Return a copy of the array ``A`` containing the same values but
      in the shape of the domain ``D``. The number of indices in the
      domain must equal the number of elements in the array. The
      elements of ``A`` are copied into the new array using the
-     default iteration orders over ``D`` and ``A``.  */
+     default iteration orders over ``D`` and ``A``.
+
+     .. note::
+
+        In addition to the above version of reshape(), there is
+        another experimental version that is available when compiling
+        with ``--edition=pre-edition`` that provides new features
+        including:
+
+        * creating a reshape() that aliases the original array rather
+          than copying it
+        * reshaping using a series of ranges rather than a domain
+        * reshaping to a 1D inferred-size array using an unbounded
+          range like ``1..``
+
+        Note that in order to capture an alias in a named variable, a
+        ``ref`` declaration must be used, like:
+
+        .. code-block:: chapel
+
+           var A = [1, 2, 3, 4];
+           ref B = reshape(A, 1..2, 1..2);
+
+        In addition, an aliasing reshape() expression can be passed as
+        a ``ref`` argument.  Aliasing reshapes are only supported for
+        local, default rectangular arrays and domains at present.  To
+        reshape other array types, see the ``copy`` argument below.
+
+        The prototypes of these new versions are essentially:
+
+        .. code-block:: chapel
+
+           proc reshape(arr: [], dom: domain(?),
+                        checkDims=checkReshapeDimsByDefault, param copy=false);
+           proc reshape(arr: [], ranges: range(?)...,
+                        checkDims=checkReshapeDimsByDefault, param copy=false);
+
+        This new version supports checking for likely mistakes in
+        which an existing dimension is split across multiple
+        dimensions as a result of the reshape.  This is controlled by
+        the ``checkDims`` argument, where ``checkReshapeDimsByDefault`` is
+        a ``config param`` whose default is ``true`` when bounds checks
+        are on and ``false`` when they're off.
+
+        The ``copy`` argument can be used to request a copy, rather than
+        an alias of the array.
+
```

Comment:
I notice this procedure doesn't use doctags to describe the formals. Should it? E.g.,

```
:arg A: The array to reshape
:type A: Any local default rectangular array
...
```

I ask because I just wrote some docs and I'm never certain of what the defaults/standards are supposed to be. Being an array type, this will appear in the spec section instead of modules docs.

And if the rest of this module doesn't follow that precedent then we probably shouldn't change it here. Just wondering what the standard is.

Reply:
I think you're right that, in general, that's a better approach.  In this case, since I'm documenting a non-default edition, I don't think it would probably work well, though?  (in that the routine I'm documenting is nodoc'd and my docs are appearing in a note environment on a different routine?).

Or wait, maybe you're suggesting I should clean up the legacy documentation?  It does appear that only two of the many routines in this file use `:arg` declarations.  And in this case, the description of reshape() refers to `A` and `D` so much in its definition that it might seem ridiculously redundant?  I think I'm going to leave the pre-existing stuff as-is, mostly due to considering it out of scope for this PR.